src/bugpoint.rs

```diff
@@ -0,0 +1,312 @@
+    RemoveEbb(Ebb),
+}
+
+enum PhaseStepResult {
+    Shrinked(String),
+    Replaced(String),
+    NoChange,
+    NextPhase(&'static str, usize),
+    Finished,
+}
+
+impl Phase {
+    fn step(&mut self, func: &mut Function) -> PhaseStepResult {
+        let first_ebb = func.layout.entry_block().unwrap();
+        match self {
+            Phase::Started => {
+                *self = Phase::RemoveInst(first_ebb, func.layout.first_inst(first_ebb).unwrap());
+                PhaseStepResult::NextPhase("remove inst", inst_count(func))
+            }
+            Phase::RemoveInst(ref mut ebb, ref mut inst) => {
+                if let Some((prev_ebb, prev_inst)) = next_inst_ret_prev(func, ebb, inst) {
+                    func.layout.remove_inst(prev_inst);
+                    if func.layout.ebb_insts(prev_ebb).next().is_none() {
+                        // Make sure empty ebbs are removed, as `next_inst_ret_prev` depends on non empty ebbs
+                        func.layout.remove_ebb(prev_ebb);
+                        PhaseStepResult::Shrinked(format!(
+                            "Remove inst {} and empty ebb {}",
+                            prev_inst, prev_ebb
+                        ))
+                    } else {
+                        PhaseStepResult::Shrinked(format!("Remove inst {}", prev_inst))
+                    }
+                } else {
+                    *self = Phase::ReplaceInstWithIconst(
+                        first_ebb,
+                        func.layout.first_inst(first_ebb).unwrap(),
+                    );
+                    PhaseStepResult::NextPhase("replace inst with iconst", inst_count(func))
+                }
+            }
+            Phase::ReplaceInstWithIconst(ref mut ebb, ref mut inst) => {
+                if let Some((_prev_ebb, prev_inst)) = next_inst_ret_prev(func, ebb, inst) {
+                    let results = func.dfg.inst_results(prev_inst);
+                    if results.len() == 1 {
+                        let ty = func.dfg.value_type(results[0]);
+                        func.dfg.replace(prev_inst).iconst(ty, 0);
+                        PhaseStepResult::Replaced(format!(
+                            "Replace inst {} with iconst.{}",
+                            prev_inst, ty
+                        ))
+                    } else {
+                        PhaseStepResult::NoChange
+                    }
+                } else {
+                    *self = Phase::ReplaceInstWithTrap(
+                        first_ebb,
+                        func.layout.first_inst(first_ebb).unwrap(),
+                    );
+                    PhaseStepResult::NextPhase("replace inst with trap", inst_count(func))
+                }
+            }
+            Phase::ReplaceInstWithTrap(ref mut ebb, ref mut inst) => {
+                if let Some((_prev_ebb, prev_inst)) = next_inst_ret_prev(func, ebb, inst) {
+                    func.dfg.replace(prev_inst).trap(TrapCode::User(0));
```

Comment:
Replacing an instruction by a trap (which is a terminal instruction) in the middle of a block means that this block has to be split, which is likely to reduce the number of verifier errors encountered while reducing. Maybe a good idea for a follow-up!